tests/unit/test_solvers.py

```diff
@@ -524,6 +524,315 @@ def test_herding_state(self, reduce_problem: _ReduceProblem) -> None:
+        evaluates to:
+
+        .. math::
+            \mathbb{E}[k(x, x')] - \frac{1}{T+1}\sum_{t=1}^T k(x, x_t) = \begin{pmatrix}
+                0.9778238600172561 - 0.4937889002469407 \\
+                0.9906914124997632 - 0.5 \\
+                0.9767967388544317 - 0.4922482185027042
+            \end{pmatrix}
+
+        giving a final result of:
+
+        .. math::
+            \mathbb{E}[k(x, x')] - \frac{1}{T+1}\sum_{t=1}^T k(x, x_t) = \begin{pmatrix}
+                0.4840349597703154 \\
+                0.4906914124997632 \\
+                0.4845485203517275
+            \end{pmatrix}
+
+        The largest value in this array is at index 1, which would be to again choose
+        the point [0.4, 0.2] for the coreset. Since we don't enforce the coreset points
+        selected to be unique here, our final ``coreset_indices`` should be [1, 1].
+
+        Finally, the solver state tracks variables we need not compute repeatedly. In
+        the case of kernel herding, we don't need to recompute
+        :math:`\mathbb{E}[k(x, x')]` at every single step - so the solver state from the
+        coreset reduce method should be set to:
+
+        .. math::
+            \mathbb{E}[k(x, x')] = \begin{pmatrix}
+                0.9778238600172561 \\
+                0.9906914124997632 \\
+                0.9767967388544317
+            \end{pmatrix}
+        """  # noqa: E501
+        # pylint: enable=line-too-long
+        # Setup example data - note we have specifically selected points that are very
+        # close to manipulate the penalty applied for nearby points, and hence enable
+        # a check of unique points using the same data.
+        coreset_size = 2
+        length_scale = 1.0 / jnp.sqrt(2)
+        x = jnp.array(
+            [
+                [0.3, 0.25],
+                [0.4, 0.2],
+                [0.5, 0.125],
+            ]
+        )
+
+        # Define a kernel
+        kernel = SquaredExponentialKernel(length_scale=length_scale)
+
+        # Generate the coreset
+        data = Data(x)
+        solver = KernelHerding(coreset_size=coreset_size, kernel=kernel, unique=False)
+        coreset, solver_state = solver.reduce(data)
+
+        # Define the expected outputs, following the arguments in the docstring
+        expected_coreset_indices = jnp.array([1, 1])
+        expected_gramian_row_mean = jnp.array(
+            [
+                0.9778238600172561,
+                0.9906914124997632,
+                0.9767967388544317,
+            ]
+        )
+
+        # Check output matches expected
+        np.testing.assert_array_equal(
+            coreset.unweighted_indices, expected_coreset_indices
+        )
+        np.testing.assert_array_equal(
+            coreset.coreset.data, data.data[expected_coreset_indices]
+        )
+        np.testing.assert_array_almost_equal(
+            solver_state.gramian_row_mean, expected_gramian_row_mean
+        )
+
 
```

Comment:
This is excellent! SonarQube might complain that some of the code is repeated/can be refactored but I am a fan of clarity.

Reply:
Added a weighted example